File: lib/world.ts
// User-built world objects. Coordinates are Z-up metres, matching the URDF / PyBullet frame:
// the robot starts at the origin facing +X. `z` is the elevation of an object's bottom face.

export type ShapeKind = 'box' | 'cylinder' | 'sphere' | 'cone' | 'pyramid' | 'wedge';

export type WorldObject = {
  id: string;
  kind: ShapeKind;
  name: string;
  x: number;
  y: number;
  z: number;
  w: number;
  d: number;
  h: number;
  yaw: number; // degrees
  color: string;
  dynamic?: boolean;
};

/** Which dimension fields a shape exposes, and how they are tied together. */
export type DimMode = 'wdh' | 'round' | 'sphere';

export type ShapeDef = {
  kind: ShapeKind;
  label: string;
  color: string;
  size: [number, number, number];
  dims: DimMode;
  preset?: string;
};

export const SHAPES: ShapeDef[] = [
  { kind: 'box', label: 'Box', color: '#e5403b', size: [1, 1, 1], dims: 'wdh' },
  { kind: 'cylinder', label: 'Cylinder', color: '#f28c1b', size: [1, 1, 1], dims: 'round' },
  { kind: 'sphere', label: 'Sphere', color: '#2f8de4', size: [1, 1, 1], dims: 'sphere' },
  { kind: 'cone', label: 'Cone', color: '#8e44c2', size: [1, 1, 1], dims: 'round' },
  { kind: 'pyramid', label: 'Pyramid', color: '#f2c21b', size: [1, 1, 1], dims: 'wdh' },
  { kind: 'wedge', label: 'Ramp', color: '#3aa655', size: [1.5, 1, 0.6], dims: 'wdh' },
  { kind: 'box', label: 'Wall', color: '#ff8a5c', size: [0.2, 3, 1], dims: 'wdh', preset: 'wall' },
];

export const dimMode = (kind: ShapeKind): DimMode =>
  kind === 'sphere' ? 'sphere' : kind === 'cylinder' || kind === 'cone' ? 'round' : 'wdh';

const SPAWN_Y = [0, 1.5, -1.5, 3, -3];

export function newObject(def: ShapeDef, existing: WorldObject[]): WorldObject {
  const n = existing.length;
  const [w, d, h] = def.size;
  const sameName = existing.filter((o) => o.name.startsWith(def.label)).length;
  return {
    id: Math.random().toString(36).slice(2, 9),
    kind: def.kind,
    name: sameName ? `${def.label} ${sameName + 1}` : def.label,
    // Drop new shapes in front of the robot, fanning out sideways so they don't overlap.
    x: 2 + Math.floor(n / SPAWN_Y.length) * 1.5,
    y: SPAWN_Y[n % SPAWN_Y.length],
    z: 0,
    w,
    d,
    h,
    yaw: 0,
    color: def.color,
  };
}

/** Apply a dimension edit while keeping round shapes round. */
export function withDims(obj: WorldObject, patch: Partial<Pick<WorldObject, 'w' | 'd' | 'h'>>): WorldObject {
  const next = { ...obj, ...patch };
  const mode = dimMode(obj.kind);
  if (mode === 'sphere') {
    const s = patch.w ?? patch.d ?? patch.h ?? obj.w;
    return { ...next, w: s, d: s, h: s };
  }
  if (mode === 'round') {
    const s = patch.w ?? patch.d ?? obj.w;
    return { ...next, w: s, d: s };
  }
  return next;
}

const CONE_SEGMENTS = 32;

/** Convex-hull vertices for cone / pyramid / wedge. Must match hull_points in backend/world.py. */
export function hullPoints(kind: ShapeKind, w: number, d: number, h: number): [number, number, number][] {
  if (kind === 'cone') {
    const ring: [number, number, number][] = [];
    for (let i = 0; i < CONE_SEGMENTS; i++) {
      const a = (2 * Math.PI * i) / CONE_SEGMENTS;
      ring.push([(w / 2) * Math.cos(a), (d / 2) * Math.sin(a), -h / 2]);
    }
    return [...ring, [0, 0, h / 2]];
  }
  if (kind === 'pyramid') {
    return [
      [-w / 2, -d / 2, -h / 2],
      [w / 2, -d / 2, -h / 2],
      [w / 2, d / 2, -h / 2],
      [-w / 2, d / 2, -h / 2],
      [0, 0, h / 2],
    ];
  }
  // wedge: ramp rising toward +x
  return [
    [-w / 2, -d / 2, -h / 2],
    [w / 2, -d / 2, -h / 2],
    [w / 2, -d / 2, h / 2],
    [-w / 2, d / 2, -h / 2],
    [w / 2, d / 2, -h / 2],
    [w / 2, d / 2, h / 2],
  ];
}

File: backend/world.py
"""User-built world objects (Tinkercad-style shapes) as static PyBullet bodies.

Sizes are full extents in metres: w along x, d along y, h along z. `z` is the
elevation of the object's bottom face. The convex-hull vertex lists here must
match `hullPoints` in lib/world.ts so the viewport shows what physics collides with.
"""

import math
from typing import Dict, List

import pybullet as p

KINDS = ("box", "cylinder", "sphere", "cone", "pyramid", "wedge")
CONE_SEGMENTS = 32


def hull_points(kind: str, w: float, d: float, h: float) -> List[List[float]]:
    if kind == "cone":
        ring = [
            [w / 2 * math.cos(2 * math.pi * i / CONE_SEGMENTS), d / 2 * math.sin(2 * math.pi * i / CONE_SEGMENTS), -h / 2]
            for i in range(CONE_SEGMENTS)
        ]
        return ring + [[0.0, 0.0, h / 2]]
    if kind == "pyramid":
        return [[-w / 2, -d / 2, -h / 2], [w / 2, -d / 2, -h / 2], [w / 2, d / 2, -h / 2], [-w / 2, d / 2, -h / 2], [0.0, 0.0, h / 2]]
    if kind == "wedge":  # ramp rising toward +x
        return [
            [-w / 2, -d / 2, -h / 2], [w / 2, -d / 2, -h / 2], [w / 2, -d / 2, h / 2],
            [-w / 2, d / 2, -h / 2], [w / 2, d / 2, -h / 2], [w / 2, d / 2, h / 2],
        ]
    raise ValueError(kind)


def volume(kind: str, w: float, d: float, h: float) -> float:
    return {
        "box": w * d * h,
        "cylinder": math.pi / 4 * w * d * h,
        "sphere": math.pi / 6 * w * w * w,
        "cone": math.pi / 12 * w * d * h,
        "pyramid": w * d * h / 3,
        "wedge": w * d * h / 2,
    }[kind]


def create_body(obj: Dict) -> int:
    kind = obj["kind"]
    w, d, h = (max(0.01, float(obj[k])) for k in ("w", "d", "h"))
    if kind == "box":
        shape = p.createCollisionShape(p.GEOM_BOX, halfExtents=[w / 2, d / 2, h / 2])
    elif kind == "cylinder":
        shape = p.createCollisionShape(p.GEOM_CYLINDER, radius=w / 2, height=h)
    elif kind == "sphere":
        shape = p.createCollisionShape(p.GEOM_SPHERE, radius=w / 2)
    elif kind in KINDS:
        shape = p.createCollisionShape(p.GEOM_MESH, vertices=hull_points(kind, w, d, h))
    else:
        raise ValueError(f"Unknown shape kind: {kind}")
    orn = p.getQuaternionFromEuler([0, 0, math.radians(float(obj.get("yaw", 0.0)))])
    pos = [float(obj["x"]), float(obj["y"]), float(obj["z"]) + h / 2]
    mass = 0.0
    if obj.get("dynamic"):
        mass = float(obj.get("mass") or min(10.0, max(0.1, volume(kind, w, d, h) * 20.0)))
    body = p.createMultiBody(baseMass=mass, baseCollisionShapeIndex=shape, basePosition=pos, baseOrientation=orn)
    if mass > 0:
        p.changeDynamics(body, -1, lateralFriction=0.8, restitution=0.5 if kind == "sphere" else 0.15)
    return body
